processor/normalizesumsprocessor/processor.go

```diff
@@ -97,229 +93,96 @@ func (nsp *NormalizeSumsProcessor) transformMetrics(rms pdata.ResourceMetrics) [
+	metricIdentifier := dataPointIdentifier(resource, metric, dp.Attributes())
 
 	start := nsp.history[metricIdentifier]
 	// If this is the first time we've observed this unique metric,
 	// record it as the start point and do not report this data point
 	if start == nil {
-		dps := metric.IntSum().DataPoints()
-		newDP := pdata.NewIntDataPoint()
-		dps.At(0).CopyTo(newDP)
+		newDP := pdata.NewNumberDataPoint()
+		dp.CopyTo(newDP)
+		newDP2 := pdata.NewNumberDataPoint()
+		newDP.CopyTo(newDP2)
 
 		newStart := startPoint{
-			dataType:     pdata.MetricDataTypeIntSum,
-			intDataPoint: &newDP,
-			lastIntValue: newDP.Value(),
+			start: newDP,
+			last:  newDP2,
 		}
 		nsp.history[metricIdentifier] = &newStart
 
-		return false
+		return
 	}
 
 	// If this data is older than the start point, we can't meaningfully report this point
 	// TODO - consider resetting on two subsequent data points older than current start timestamp.
 	// This could signify a permanent clock change.
-	if dp.Timestamp() <= start.intDataPoint.Timestamp() {
+	if dp.Timestamp() <= start.start.Timestamp() {
 		nsp.logger.Info(
 			"data point being processed older than last recorded reset, will not be emitted",
-			zap.String("lastRecordedReset", start.intDataPoint.Timestamp().String()),
+			zap.String("lastRecordedReset", start.start.Timestamp().String()),
 			zap.String("dataPoint", dp.Timestamp().String()),
 		)
-		return false
+		return
 	}
 
 	// If data has rolled over or the counter has been restarted for
 	// any other reason, grab a new start point and do not report this data
-	if dp.Value() < start.lastIntValue {
-		dp.CopyTo(*start.intDataPoint)
-		start.lastIntValue = dp.Value()
+	if (dp.Type() == pdata.MetricValueTypeInt && dp.IntVal() < start.last.IntVal()) || (dp.Type() == pdata.MetricValueTypeDouble && dp.DoubleVal() < start.last.DoubleVal()) {
+		dp.CopyTo(start.start)
+		dp.CopyTo(start.last)
 
-		return false
+		return
 	}
 
-	start.lastIntValue = dp.Value()
-	dp.SetValue(dp.Value() - start.intDataPoint.Value())
-	dp.SetStartTimestamp(start.intDataPoint.Timestamp())
+	dp.CopyTo(start.last)
 
-	return true
+	newDP := ndps.AppendEmpty()
+	dp.CopyTo(newDP)
+	switch dp.Type() {
+	case pdata.MetricValueTypeInt:
+		newDP.SetIntVal(dp.IntVal() - start.start.IntVal())
+	case pdata.MetricValueTypeDouble:
+		newDP.SetDoubleVal(dp.DoubleVal() - start.start.DoubleVal())
+	}
+	newDP.SetStartTimestamp(start.start.Timestamp())
 }
 
-func dataPointIdentifier(resource pdata.Resource, metric pdata.Metric, labels pdata.StringMap) string {
+func dataPointIdentifier(resource pdata.Resource, metric pdata.Metric, labels pdata.AttributeMap) string {
 	var b strings.Builder
 
 	// Resource identifiers
 	resource.Attributes().Sort().Range(func(k string, v pdata.AttributeValue) bool {
-		fmt.Fprintf(&b, "%s=", k)
-		addAttributeToIdentityBuilder(&b, v)
-		b.WriteString("|")
+		fmt.Fprintf(&b, "%s=%s|", k, v.AsString())
```

Comment:
The only potential issue with v.AsString() is (last I looked) it doesn't sort map type attribute values, which could lead to inconsistent data point identifiers. This is pretty edge case, but when I was looking at refactoring normalizesums for 0.36 I noticed this. Should we be concerned about this edge case?

Reply:
Nevermind, it looks like it's using json.marshal which should produce consistent order. 👍